# Qustion 1: Vacation Planner

### Single Point Algorithm
For each sequence pair selected, a location is chosen from the genes in between, except for the first and last gene. This is the crossover point. The genes that come after this point are mutually displaced in both sequences. For this operation, the arrays must be of the same length.

In [31]:
from random import randint
import random

def individual(length, min, max):
    return [randint(min,max) for x in range(length)]
 
def population(count, length, min, max):
    return [individual(length, min, max) for x in range(count)]

def fitness(individual, target):
    total = sum(individual)
    return abs(target-total)

def grade(pop, target):
    summed = [sum(i)-target for i in pop]
    return (sum(summed) / len(pop))

def evolve(pop, target, retain=0.2, random_select=0.05, mutate=0.01):
    graded = [(fitness(x, target), x) for x in pop]
    graded = [x[1] for x in sorted(graded)] # x[1] because x has two components, just take the list
    retain_length = int(len(graded)*retain) # how many top % parents to be remained
    parents = graded[0:retain_length] # get list of array of individuals as parents - after sorted
    
    # randomly add other individuals to promote genetic diversity
    for individual in graded[retain_length:]:
        if random_select > random.random():
            parents.append(individual)
    
    # mutate some individuals
    for individuals in parents:
        if mutate > random.random():
            pos_to_mutate = randint(0, len(individual)-1)
            individual[pos_to_mutate] = randint(min(individual), max(individual))
    
    # crossover parents to create children
    parents_length = len(parents)
    desired_length = len(pop) - parents_length # to make sure have enough individuals
    children = []
    while len(children) < desired_length:
        male = randint(0, parents_length-1)
        female = randint(0, parents_length-1)
        if male != female:
            male = parents[male]
            female = parents[female]
            half = int(len(male) / 2)
            child = male[:half] + female[half:]
            children.append(child)
            
    parents.extend(children)
    return parents

In [32]:
value_lst =[]
fitness_history = []

target = 5000
p_count = 100
i_length = 4
i_min = 1
i_max = 1500
n_generation = 100

p = population(p_count, i_length, i_min, i_max)

for i in range(n_generation):
    p = evolve(p, target)
    value = grade(p, target)
    fitness_history.append(value)
    value_lst.append(p[0])
    value_lst.append(value)
    

for datum in fitness_history:
    print(datum)
    
value_lst

-1133.2
-301.36
-76.68
87.28
-40.12
-9.32
3.24
-2.3
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-0.92
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
4.76
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0


[[1004, 1098, 1436, 1478],
 -1133.2,
 [1004, 1098, 1436, 1478],
 -301.36,
 [1144, 1474, 1131, 1249],
 -76.68,
 [1144, 1474, 1131, 1249],
 87.28,
 [1144, 1474, 1131, 1249],
 -40.12,
 [1144, 1474, 1131, 1249],
 -9.32,
 [1144, 1474, 1131, 1249],
 3.24,
 [1144, 1474, 1131, 1249],
 -2.3,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -0.92,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 [1144, 1474, 1131, 1249],
 -2.0,
 

[1144, 1474, 1131, 1249]

Scenario: <br/>
Money on hand = RM 5000 <br/>
Vacation duration = 5 <br/>
Hotel Star rating = 1144/5 = RM228.8 per night <br/>
Tourist spots = 6 spots <br/>
One tourist spot = 1474/6 = RM245.67 <br/>
Food price = 1131/15 = RM75.4 per meal <br/>
Transport frequency = 10 trip per day <br/>
Trip per day = 1249/50 = RM24.98 per trip <br/>

Money Saved = RM2

### Two Point Algorithm
In this method, two point are selected on the sequence, except for the first and last genes. The crossover process is the displacement of genes between these two selected points and genes are exchanged between these points.

In [33]:
from random import randint
import random

def individual(length, min, max):
    return [randint(min,max) for x in range(length)]
 
def population(count, length, min, max):
    return [individual(length, min, max) for x in range(count)]

def fitness(individual, target):
    total = sum(individual)
    return abs(target-total)

def grade(pop, target):
    summed = [sum(i)-target for i in pop]
    return (sum(summed) / len(pop))

def evolve(pop, target, retain=0.2, random_select=0.1, mutate=0.1):
    graded = [(fitness(x, target), x) for x in pop]
    graded = [x[1] for x in sorted(graded)] # x[1] because x has two components, just take the list
    retain_length = int(len(graded)*retain) # how many top % parents to be remained
    parents = graded[0:retain_length] # get list of array of individuals as parents - after sorted
    
    # randomly add other individuals to promote genetic diversity
    for individual in graded[retain_length:]:
        if random_select > random.random():
            parents.append(individual)
    
    # mutate some individuals
    for individuals in parents:
        if mutate > random.random():
            pos_to_mutate = randint(0, len(individual)-1)
            individual[pos_to_mutate] = randint(min(individual), max(individual))
    
    # crossover parents to create children
    parents_length = len(parents)
    desired_length = len(pop) - parents_length # to make sure have enough individuals
    children = []
    while len(children) < desired_length:
        male = randint(0, parents_length-1)
        female = randint(0, parents_length-1)
        if male != female:
            male = parents[male]
            female = parents[female]
            child = male[:2] + female[2:4] + male[4:]
            children.append(child)
            
    parents.extend(children)
    return parents

In [34]:
value_lst =[]
fitness_history = []

target = 5000
p_count = 100
i_length = 4
i_min = 1
i_max = 1500
n_generation = 100

p = population(p_count, i_length, i_min, i_max)

for i in range(n_generation):
    p = evolve(p, target)
    value = grade(p, target)
    fitness_history.append(value)
    value_lst.append(p[0])
    value_lst.append(value)
    

for datum in fitness_history:
    print(datum)
    
value_lst

-1118.75
-560.79
-81.39
-102.7
-67.31
29.78
2.91
-23.15
1.47
-3.0
-3.0
-3.0
-3.0
-1.77
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-12.26
-6.92
-3.0
-3.0
-3.0
-3.0
-9.98
-3.0
-3.0
-3.0
-3.0
2.52
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-6.45
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
-3.0
0.7
8.13
0.7
-2.16
-2.03
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0
-2.0


[[1306, 1111, 985, 1358],
 -1118.75,
 [1215, 1297, 1132, 1353],
 -560.79,
 [1215, 1297, 1132, 1353],
 -81.39,
 [1215, 1297, 1132, 1353],
 -102.7,
 [1215, 1297, 1132, 1353],
 -67.31,
 [1215, 1297, 1132, 1353],
 29.78,
 [1215, 1297, 1132, 1353],
 2.91,
 [1215, 1297, 1132, 1353],
 -23.15,
 [1215, 1297, 1132, 1353],
 1.47,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -1.77,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0,
 [1215, 1297, 1132, 1353],
 -3.0

[1216, 1297, 1132, 1353]

Scenario: <br/>
Money on hand = RM 5000 <br/>
Vacation duration = 5 <br/>
Hotel Star rating = 1216/5 = RM243.2 per night <br/>
Tourist spots = 6 spots <br/>
One tourist spot = 1297/6 = RM216.17 <br/>
Food price = 1132/15 = RM75.47 per meal <br/>
Transport frequency = 10 trip per day <br/>
Trip per day = 1353/50 = RM27.06 per trip <br/>

Money Saved = RM2

### Uniform Algorithm
Changes are made between randomly selected pairs of chromosomes using a probability value.

In [29]:
from random import randint
import random

def individual(length, min, max):
    return [randint(min,max) for x in range(length)]
 
def population(count, length, min, max):
    return [individual(length, min, max) for x in range(count)]

def fitness(individual, target):
    total = sum(individual)
    return abs(target-total)

def grade(pop, target):
    summed = [sum(i)-target for i in pop]
    return (sum(summed) / len(pop))

def evolve(pop, target, retain=0.2, random_select=0.25, mutate=0.25):
    graded = [(fitness(x, target), x) for x in pop]
    graded = [x[1] for x in sorted(graded)] # x[1] because x has two components, just take the list
    retain_length = int(len(graded)*retain) # how many top % parents to be remained
    parents = graded[0:retain_length] # get list of array of individuals as parents - after sorted
    
    # randomly add other individuals to promote genetic diversity
    for individual in graded[retain_length:]:
        if random_select > random.random():
            parents.append(individual)
    
    # mutate some individuals
    for individuals in parents:
        if mutate > random.random():
            pos_to_mutate = randint(0, len(individual)-1)
            individual[pos_to_mutate] = randint(min(individual), max(individual))
    
    # crossover parents to create children
    parents_length = len(parents)
    desired_length = len(pop) - parents_length # to make sure have enough individuals
    children = []
    crossprob = 0.5
    while len(children) < desired_length:
        child = []
        male = randint(0, parents_length-1)
        female = randint(0, parents_length-1)
        if male != female:
            male = parents[male]
            female = parents[female]
            Leng = int(len(male))
            for i in range(Leng):
                if random.random() < crossprob:
                    temp = male[i]
                    child.append(temp)
                else:
                    temp = female[i]
                    child.append(temp) 
                    
            children.append(child)
    parents.extend(children)
    return parents

In [30]:
value_lst =[]
fitness_history = []

target = 5000
p_count = 100
i_length = 4
i_min = 1
i_max = 1500
n_generation = 100

p = population(p_count, i_length, i_min, i_max)

for i in range(n_generation):
    p = evolve(p, target)
    value = grade(p, target)
    fitness_history.append(value)
    value_lst.append(p[0])
    value_lst.append(value)
    

for datum in fitness_history:
    print(datum)
    
value_lst

-1620.16
-1397.17
-1085.91
-720.54
-633.73
-470.77
-164.01
32.28
-10.58
-1.86
10.12
9.46
-12.72
-10.47
-24.47
-24.23
-11.37
-3.52
-4.33
-5.38
-5.42
0.3
-2.32
-3.44
3.53
-4.0
-4.52
-2.92
0.53
-4.0
-4.0
-4.0
-4.0
-4.0
-4.0
-4.0
-4.0
-4.0
-4.0
-6.6
-4.0
-4.0
-5.93
-4.0
-4.0
-4.0
-4.0
-5.26
-14.46
-4.0
-4.0
2.65
-4.0
-4.0
-4.0
-4.34
-4.0
-4.0
-4.0
-4.0
-4.0
-4.0
-4.0
-4.0
-4.0
-4.0
-4.0
-4.0
-0.16
13.38
-3.92
-2.81
-2.41
2.58
-1.73
-6.41
1.8
-15.63
0.7
-9.13
8.24
3.09
-1.53
0.0
0.0
0.0
0.47
0.0
0.0
0.0
0.0
0.0
-10.97
-0.3
-0.24
0.0
0.0
0.0
0.0
0.0


[[1200, 1040, 1410, 1412],
 -1620.16,
 [1023, 1420, 1486, 1127],
 -1397.17,
 [1200, 1420, 1136, 1207],
 -1085.91,
 [1200, 1420, 1149, 1207],
 -720.54,
 [1256, 1353, 1410, 976],
 -633.73,
 [1256, 1353, 1410, 976],
 -470.77,
 [1200, 1420, 1169, 1207],
 -164.01,
 [1200, 1420, 1169, 1207],
 32.28,
 [1200, 1420, 1169, 1207],
 -10.58,
 [1200, 1420, 1169, 1207],
 -1.86,
 [1200, 1420, 1169, 1207],
 10.12,
 [1200, 1420, 1169, 1207],
 9.46,
 [1200, 1420, 1169, 1207],
 -12.72,
 [1200, 1420, 1169, 1207],
 -10.47,
 [1200, 1420, 1169, 1207],
 -24.47,
 [1200, 1420, 1169, 1207],
 -24.23,
 [1200, 1420, 1169, 1207],
 -11.37,
 [1200, 1420, 1169, 1207],
 -3.52,
 [1200, 1420, 1169, 1207],
 -4.33,
 [1200, 1420, 1169, 1207],
 -5.38,
 [1200, 1420, 1169, 1207],
 -5.42,
 [1200, 1420, 1169, 1207],
 0.3,
 [1200, 1420, 1169, 1207],
 -2.32,
 [1200, 1420, 1169, 1207],
 -3.44,
 [1200, 1420, 1169, 1207],
 3.53,
 [1200, 1420, 1169, 1207],
 -4.0,
 [1200, 1420, 1169, 1207],
 -4.52,
 [1200, 1420, 1169, 1207],
 -2.92,
 [12

[1200, 1317, 1169, 1314]

Scenario: <br/>
Money on hand = RM 5000 <br/>
Vacation duration = 5 <br/>
Hotel Star rating = 1200/5 = RM240 per night <br/>
Tourist spots = 6 spots <br/>
One tourist spot = 1317/6 = RM219.5 <br/>
Food price = 1169/15 = RM77.93 per meal <br/>
Transport frequency = 10 trip per day <br/>
Trip per day = 1314/50 = RM26.28 per trip <br/>

Money Saved = RM0

In [1]:
from tabulate import tabulate

mydata = [
    ["Single Point Algorithm", 2],
    ["Two Point Algorithm", 2],
    ["Uniform Algorithm", 0]
]
    
head = ["Algorithm Type", "Money Saved"]
print(tabulate(mydata, headers=head, tablefmt="grid"))

+------------------------+---------------+
| Algorithm Type         |   Money Saved |
+========================+===============+
| Single Point Algorithm |             2 |
+------------------------+---------------+
| Two Point Algorithm    |             2 |
+------------------------+---------------+
| Uniform Algorithm      |             0 |
+------------------------+---------------+
